### DISCUSSION TITLE
[SUGGESTION] Large Scale Macro/Keybinding Manager -> Vim for Musescore?

### DISCUSSION THREAD

**Discussion by playerpiano05** [musescore/MuseScore#25929] [Ideas]
The amount of mousing that has to be done to access information in pallets is absurd. I feel it could be completely reasonable to be able to toggle a Vim mode kind of thing, say activate a selector layer with some control/cmd combo, and then each inspector pallet can be assigned a keystroke within that selector layer, and then that unlocks an action layer in itself to access items within the pallets. What this can end up doing is reducing any desired pallet item to a maximum of 4 keystrokes, and ease basically everyone's concerns about being unable to access things with the keyboard. Thoughts?

**Comment by playerpiano05** [comment]
this could also be an overall (or at least optional overall; toggleable?) to how navigation is done throughout the entire app. Think normal mode: moving through individual notes with h/l  moving up and down across lines using j/k, move through measures (words) with w/e/b, etc. for those not familiar with vim motions, look here: https://vim.rtorr.com/. note input mode can be engaged with i as per usual, and select mode can be accessed by v, just as in vim. Of course, these keys can be changed. there should be a dialog for a user to configure their entire keyboard if they so choose, but i think that there should be a nonzero overlap of vim users and musicians (since I am here, at least). making the entire engraving process mouse-free would be a HUGE improvement for musescore and how it's accepted, and i feel that a new control scheme is needed. the current one is clunky, there is only intuition for the most basic of the control schemes, whereas with this, there would be an underlying logic for how the scheme works that's been well understood for decades, just applied in a different way.

**Comment by MarcSabatella** [comment]
MuseScore already supports complete keyboard navigation of the UI and the score itself - it is in fact the most keyboard-accessible music notation program out there, and the top choice of blind musicians for exactly this reason.  Be sure to read the Handbook section on keyboard shortcuts and on accessibility to learn how to do this (eg, Left/Right plus Ctrl and/or Alt to navigate the score horizontally, Alt+Up/Down vertically, Tab, F6, and cursor to navigate the UI, etc), and ask questions on the Support forum at musescore.org, if you have questions.

That said, nothing wrong with the original request for even better keyboard navigation of the palettes, and this is already on the radar.

**Reply by playerpiano05** [reply]
I'm definitely well aware that the current navigation is effective; I was mainly suggesting having a separate "VIM mode" to maintain logical consistency, and allow for easy transfer of familiarity. Do you think it would be reasonable for me to develop both of these features to be toggled? Or do you think only building the navigation of pallets would be good? My thought is that a rewrite of the control surface (as a toggleable mode) could also really help with percussion note input, as entry exceeds the notes A-G, and using the on screen keyboard to access additional notes is cumbersome. Thank you for engaging with this idea, and let me know your further thoughts.

**Reply by myrmeleontidae** [reply]
Sorry to necro-bump a ten-month-old post, but I wholeheartedly agree with your suggestions and found this thread in the first place because the currently implemented keyboard shortcut assignment menu in MuseScore leaves a lot to be desired. That, and I grew weary of editing someone else's attempt at a very loosely Vim-inspired .xml shortcut file upon the realization that he'd used accented characters from a French keyboard for a few of the bindings.

Considering MuseScore uses Qt, it might be worth reaching out to @The-Compiler, qutebrowser's developer. If you're unfamiliar, it's a web browser that, while maintaining a Vim-like feature set, also supports hinting for the selection of GUI elements and is mostly written in Python. It's made internet browsing on my laptop on-the-go an absolute pleasure instead of a chore. Given the nature of score writer / MIDI editor software, numerous graphical elements are kind of unavoidable, but something like qutebrowser's hinting layer would do wonders for ease of use, at least for me; trackpads suck, haha. To the point where I have half a mind to go digging into the docs for MuseScore plugin implementation.

**Comment by playerpiano05** [comment]
The reason why I'm such a big proponent of vim motions is that it can massively simplify selection. Instead of having to delete -> shift arrow over two measures and five notes ,(9 key presses with time delay for visual feedback, room for error) one could literally just type d2w5k to get a precise selection and deletion.

**Comment by MarcSabatella** [comment]
You don’t need a whole separate mode to change the shortcuts to H/J/K/L - just do that yourself in Edit / Preferences / Shortcuts.  VIM is probably familiar to only 0.1% of MuseScore users, so overall mimicking its quirky choices of shortcuts isn’t going to have wide appeal, but the shortcut customization facility is there for you.

Palette navigation is definitely more keystrokes than it needs to be, and there are already tentative plans on how to improve that.  Percussion input is also being completely revamped for 4.5.

**Reply by ghost** [reply]
This is missing a critical point of Vim motions.

The thing about Vim motions is that you can specify a count for a command (how many times you want to repeat it). For example, if you wanted to move 8 characters (or notes) forward, instead of pressing L 8 times or pressing the key down until you've moved 8 characters, you can type "8l". If you wanted to delete two words (or measures) while your cursor is in the middle of a word, you can type "d2aw".

You can also add motions to commands, which specify a "range" for the command. For example, "dd" deletes a line. "yy" copies or "yanks" a line. "c$" deletes all characters from the cursor to the end of the line.

The fact of the matter is that Vim's system cannot be easily replicated by just changing some shortcuts. It's a delicate system that offers blazing speed, and it wouldn't be wise to shrug it off.

**Reply by playerpiano05** [reply]
would you like to write a vim plugin with me for musescore?

**Reply by myrmeleontidae** [reply]
> would you like to write a vim plugin with me for musescore?

I'd be willing to give it a shot, so long as there's no particular deadline, haha. I only really have experience with C (and a teensy bit of assembly; embedded stuff/ I'm an electronic eng. student), but I do have the benefit of knowing what Vim motions _should_ do. Many hands make light work, or something. 

I'd love to get something like that up and running. My hands only really enjoy speaking Vim now, haha, so that kind of productivity boost for my (musical) composing would be quite welcome.

**Comment by playerpiano05** [comment]
I am in full agreement!

**Comment by ghost** [comment]
I just started using LilyPond instead of MuseScore. The best way to get a Vim mode is just by using (Neo)Vim.
